### ISSUE TITLE
Ensure that login, registration and recovery do not leak identity information (Account Enumeration Attack)

### ISSUE THREAD

**Issue by aeneasr** [ory/kratos#133]
**Is your feature request related to a problem? Please describe.**

Ensure that it is not possible to see if a user is signed up on Kratos using password reset flow.

**Context**

![](https://www.troyhunt.com/content/images/2016/02/21782177Password-Reset5.png)

**Comment by aeneasr** [comment]
This needs to be done for

- [ ] Sign up (send email to recipient and show success message)
- [ ] Log in (just show "invalid credentials")
- [ ] Account recovery (send email to recipient and show success message)

**Comment by aeneasr** [comment]
Pretty hard to implement for settings flow, pushing way back.

**Comment by mattbonnell** [comment]
Taking inventory here:
Account enumeration via
* Login flow should be protected by #1088 
* Recovery flow seems to already be protected; the user sees the same UI whether the account exists or not. (question though, if a recovery link is requested to an email that isn't associated to an account, why do we need to send them an email? Seems like an easy way to spam people at the Kratos' admin's reputational and financial expense).

This leaves the registration flow. The only way I can see to protect against account enumeration there would be to enforce an email verification step (or make it configurable through a PostRegistrationPrePersist hook) so that we can show the same UI whether an account exists for that email or not. If it doesn't, we send the verification email. If it does, we send a "You already have an account" email with a password recovery link. What are your thoughts there?

**Comment by zepatrik** [comment]
You are absolutely right, thanks for summing up the current state. The problem remaining is how to we prevent people to change the email address to an existing one without letting them know that it exists already.

**Comment by gitn00b1337** [comment]
Hello.

I'm getting "An account with the same identifier (email, phone, username, ...) exists already." when trying to register an account. I keep seeing references in github to this being fixed against account enumeration but unsure if it's actually in the latest docker image yet? Is there a way to override system messages?

Thanks!

**Comment by alantbarlow** [comment]
What is the current state of this idea?

I was thinking of a registration flow that could make this work. It basically starts submitting your email for verification before selecting the login method(s) you'd like to use. if you are already registered it will ask you to login instead. if you are not registered it will provide a verification code. Once you entered your verification code, you can then select your login method(s). 

This would completely solve this issue for registration because it will only ever ask for and send an email address to begin the flow. This is also not the same as #2029 because the email won't be a login method. It was only used to verify the users email address before signing them up.

**Comment by alantbarlow** [comment]
To solve the problem of the user changing their email and giving errors. when the email address gets changed, a verification email should be sent out to the email address asking them to verify the change. if the new email wasn't already in use, then the email would give them the verification code/magic-link. Once the email address has been verified, only then do you persist the change.

 If the chosen email address has an account already, say that there was an error changing your email address since it was linked to another account. 

We can even require email verification of both the existing and new email addresses as an extra security measure preventing account takeover.

**Comment by mitar** [comment]
> It basically starts submitting your email for verification before selecting the login method(s) you'd like to use.

Just to be clear if I understand this correctly. You are saying that Kratos would always first send an e-mail to verify the e-mail address, and only then user would select the login method? Then this is first factor passwordless authentication and second factor then user can choose from login methods?

Downside of this is that user have to do the whole email verification flow every time, even if they might have otherwise strong authentications in place (e.g., strong password and strong 2FA). And also, it does not work when there is no e-mail associated with the account (e.g., when you have just username and password authentication, i.e., for e-mail providers themselves).

**Comment by mitar** [comment]
My current take on addressing account enumeration for password-based accounts is to in fact combine login and registration flows into one flow. This is similar to how already Social Sign-in works. So then the combined login and registration flow is always:

1. enter e-mail address or phone number
2. enter password
3. enter code you got to your e-mail address or phone number

The order is important (otherwise you have passwordless authentication, and here I am focusing on password-based authentication). And this is why this works:

**Account does not yet exist**: After code is entered (thus e-mail address or phone number is verified), you create an account with given password. Even if the user mistyped password, no worries, you verified the e-mail address so they can recover (see below). Asking for profile information is done after registration (name, etc.). User is logged in after registration immediately (but this could also be allowed to be turned off, I just think it makes no sense to require to repeat literally the same flow once more). If attacker is trying to enumerate accounts, they do not learn anything because they do not have access to target e-mail accounts. They do spam them, but that is covered in #1835 already.

**Account does exist but password is invalid**: Maybe user forgot that they have an account. Maybe user just mistyped the password. Maybe this is an attacker trying to gain access. If it is an attacker, they do not have access to the e-mail address, so nothing happens, and they learn nothing. If it is user, they are not asked "have you forgotten the password", they just immediately go into password recovery and after entering the code (and doing any other MFA), they are logged in and can be prompted with message that their password was invalid and if they want to re-set it.

**Account does exist and password is correct**: We can then skip step 3., we know the user. And can simply do regular MFA for the account. If this happens to be sending a code, then also OK.

This does not work with username+password-only accounts. For them there is no step 3. For them we have to return an error message "invalid password for username" (which also means that username is already taken) after step 2. But I find that not as problematic because users can generally pick an arbitrary username while e-mail addresses are often tied to their identity.

**Comment by alantbarlow** [comment]
> Then this is first factor passwordless authentication and second factor then user can choose from login methods?

This is not what I was talking about. Passwordless authentication might be how we can verify the user's email upon registration if needed but that will not be the login method. Essentially the registration flow would look like this...

1. the user submits their email address,
2. the user enters their verification code or clicks on a magic link that was emailed to them
3. the user selects their primary login method

If you are already registered, it will give an error in the email saying so and ask you to log in instead. Then the login flow would be normal, complete only their preferred login method. The selected login method is not the second factor. 

One way this could work is to first sign every user up using passwordless to verify their email during registration but flag the account as not completed registration. Then when they complete the registration flow we update their preferred login method to something else. They could select any of the options available including passwordless, federated login, or password. Choosing one would actually change their primary login method.

**Comment by alantbarlow** [comment]
We could accomplish this with account linking, allowing them to add multiple login methods to their account with the option to disable the passwordless login that is enabled by default.

**Comment by Sytten** [comment]
Not stale, again please remove this bot

**Comment by robsoncombr** [comment]
Isn´t supposed to simply return that an e-mail was sent to the provided e-mail, phone, whatever?
Maybe for a more advanced approach, create some sort of rate block or warning for admins.

**Comment by McKean** [comment]
echoing what @robsoncombr mentioned: on sign up if an existing email is found, the same message should be shown (with code challenge). However no emails should be sent. For the session hook this gets a bit tricky as the user should not be logged in, perhaps ignoring the session hook or any other hook would be the right thing to do.

**Comment by Sytten** [comment]
Not stale
